config/crd/bases/batch.volcano.sh_jobs.yaml

```diff
@@ -41,6 +41,9 @@ spec:
           description: Specification of the desired behavior of the volcano job, including
             the minAvailable
           properties:
+            keepSchedule:
```

Comment:
Can you describe your usage scene?

Reply:
For example, in my namespace foo, there is a resource quota for 5 cpus.

```yaml
  apiVersion: v1
  kind: ResourceQuota
  metadata:
    name: quota-cpu
    namespace: foo
  spec:
    hard:
      cpu: "5"
```

And my mpijob just need 5 cpu, a mpi-launcher need 1 cpu, 2 mpi-workers need 2 cpu separately. Everything looks good.
But before I start my mpi job, I start a pod that consumes 2 cpu.

Now I just have 2 mpi pods, a launcher and a worker, I have to wait for pod completes and releases resources.
But now, job controller reaches 15 retries, and drops this req from worker's queue.
I can't get the last pod any more, even the former pod releases resources.
And the job also not releases the resources it holds.

Reply:
> For example, in my namespace foo, there is a resource quota for 5 cpus.
> 
> ```yaml
>   apiVersion: v1
>   kind: ResourceQuota
>   metadata:
>     name: quota-cpu
>     namespace: foo
>   spec:
>     hard:
>       cpu: "5"
> ```
> 
> And my mpijob just need 5 cpu, a mpi-launcher need 1 cpu, 2 mpi-workers need 2 cpu separately. Everything looks good.
> But before I start my mpi job, I start a pod that consumes 2 cpu.
> 
> Now I just have 2 mpi pods, a launcher and a worker, I have to wait for pod completes and releases resources.
> But now, job controller reaches 15 retries, and drops this req from worker's queue.
> I can't get the last pod any more, even the former pod releases resources.
> And the job also not releases the resources it holds.

Perhaps it seems reasonable in some scenes as you describe. But it may cause dead lock. For example, there are 10 cpus in the cluster. And another end-user launch a job which is same with your condition. It happens that he is also running part of his tasks and waiting your running tasks to release resouce. If you both can be wait for each other forever, your jobs both will be unfinished forever. Scheduler must take some actions to release some resources to resolve the problem. That's the reason why we set 15 tries.

Reply:
> If you both can be wait for each other forever, your jobs both will be unfinished forever. Scheduler must take some actions to release some resources to resolve the problem.

Yeah, you are right, this would lead to dead lock, maybe I should resolve this problem from the perspect of "How to release resources after controller retries 15 times".

But now, job controller just drop job out of  the worker's queue and do nothing. 
In the above case:

1. cpus are allocated, but no job works;
2. controller drops the `sync request` out of the queue, IMO job's status can not be synced;
3. cpus only could be released until job is deleted.

How about releasing job's resources after retry 15 times?
Every `xxxState` should `Execute(TerminateJobAction)` after failed times.

Reply:
> > If you both can be wait for each other forever, your jobs both will be unfinished forever. Scheduler must take some actions to release some resources to resolve the problem.
> 
> Yeah, you are right, this would lead to dead lock, maybe I should resolve this problem from the perspect of "How to release resources after controller retries 15 times".
> 
> But now, job controller just drop job out of the worker's queue and do nothing.
> In the above case:
> 
> 1. cpus are allocated, but no job works;
> 2. controller drops the `sync request` out of the queue, IMO job's status can not be synced;
> 3. cpus only could be released until job is deleted.
> 
> How about releasing job's resources after retry 15 times?
> Every `xxxState` should `Execute(TerminateJobAction)` after failed times.

Please show me your scheduler configuration. I guess you have not configed Gang plugin.

Reply:
this is my scheduler configuration

```yaml
apiVersion: v1
data:
  volcano-scheduler.conf: |
    actions: "enqueue, allocate, backfill"
    tiers:
    - plugins:
      - name: priority
      - name: gang
      - name: conformance
    - plugins:
      - name: drf
      - name: predicates
      - name: proportion
      - name: nodeorder
      - name: binpack
kind: ConfigMap
metadata:
  name: volcano-scheduler-configmap
  namespace: volcano-system
```

Reply:
> this is my scheduler configuration
> 
> ```yaml
> apiVersion: v1
> data:
>   volcano-scheduler.conf: |
>     actions: "enqueue, allocate, backfill"
>     tiers:
>     - plugins:
>       - name: priority
>       - name: gang
>       - name: conformance
>     - plugins:
>       - name: drf
>       - name: predicates
>       - name: proportion
>       - name: nodeorder
>       - name: binpack
> kind: ConfigMap
> metadata:
>   name: volcano-scheduler-configmap
>   namespace: volcano-system
> ```

How do you prove the behavior above? As i know, if gang plugin is configured, only allocatable resource satisfies at leaset minAvailable tasks can the job be running. Namly, minAvailable tasks will be allocated resource at the same time.

Reply:
@Thor-wl  Sorry, it took so long to reply.

We have deployed the a new environment with our customed scheduler. Is it useful that I describe the situation I met before?

The follow is my node's capacity, cpu are sufficient.

```log
Capacity:
 cpu:                                 48
 ephemeral-storage:                   2228605748Ki
 hugepages-1Gi:                       0
 hugepages-2Mi:                       0
 memory:                              263602412Ki
 pods:                                500
```

I want to test the behavior of scheduler with k8s resource quota, so I make a namespace that just be allocated 5 cpus.

```yaml
apiVersion: v1
kind: List
items:
- apiVersion: v1
  kind: ResourceQuota
  metadata:
    name: test-quota-cpu
    namespace: foo
  spec:
    hard:
      cpu: "5"
---
apiVersion: v1
kind: Pod
metadata:
  name: consume-2-cpu
  namespace: foo
spec:
  containers:
  - name: test
    image: busybox:latest
    command: ["/bin/sh", "-c", "tail -f /dev/null"]
    resources:
      requests:
        cpu: 2
      limits:
        cpu: 2
---
apiVersion: batch.volcano.sh/v1alpha1
kind: Job
metadata:
  name: big-cpu-mpi-job
  namespace: foo
spec:
  minAvailable: 3
  schedulerName: volcano
  keepSchedule: true
  plugins:
    ssh: []
    svc: []
  tasks:
    - replicas: 1
      name: mpimaster
      policies:
        - event: TaskCompleted
          action: CompleteJob
      template:
        spec:
          containers:
            - command:
                - /bin/sh
                - -c
                - |
                  MPI_HOST=`cat /etc/volcano/mpiworker.host | tr "\n" ","`;
                  mkdir -p /var/run/sshd; /usr/sbin/sshd;
                  mpiexec --allow-run-as-root --host ${MPI_HOST} -np 2 mpi_hello_world > /home/re;
              image: example-mpi:0.0.1
              name: mpimaster
              ports:
                - containerPort: 22
                  name: mpijob-port
              workingDir: /home
              resources:
                limits:
                  cpu: 1
          restartPolicy: OnFailure
    - replicas: 2
      name: mpiworker
      template:
        spec:
          containers:
            - command:
                - /bin/sh
                - -c
                - |
                  mkdir -p /var/run/sshd; /usr/sbin/sshd -D;
              image: example-mpi:0.0.1
              name: mpiworker
              ports:
                - containerPort: 22
                  name: mpijob-port
              workingDir: /home
              resources:
                limits:
                  cpu: 2
          restartPolicy: OnFailure
```

I guess volcano scheduler maybe just consider node's resource and ignores namespace's resource quota.

Reply:
Yeah, so Volcano should consider resource quota. It's an import point indeed.

Reply:
> Yeah, so Volcano should consider resource quota. It's an import point indeed.

Yeah, from this perspective, this is pr is a workaround to deal with this issue, I find pr https://github.com/volcano-sh/volcano/pull/1100 considers resource quota.